### ISSUE TITLE
Less fails to parse valid Custom Property values (ex: @apply polyfill)

### ISSUE THREAD

**Issue by Jamtis** [less/less.js#2715]
This code breaks:

```
paper-drawer-panel {
  --paper-drawer-panel-left-drawer: {
    background-color: red;
  };
}
```

The curly brackets in the property value break the parser.

I'd like LESS to support the `@apply` syntax, as used by Polymer.
https://www.polymer-project.org/1.0/docs/devguide/styling.html#custom-css-mixins

Also tried:

```
paper-drawer-panel {
  @ruleset: {
    prop: value;
  };
  --paper-drawer-panel-left-drawer: %(~"%d", @ruleset);
}
```

It didn't break the parser but delivered:

```
paper-drawer-panel {
  --paper-drawer-panel-left-drawer: ;
}
```

which is clearly not a desirable behavior.

**Comment by seven-phases-max** [comment]
I'm afraid supporting non-CSS syntax support is beyond Less scope. So my -1.

Either way notice you can emit whatever code into the output with escaping. E.g.:
```less
xpaper-drawer-panel {
  --paper-drawer-panel-left-drawer: ~"{background-color: red;}";
}
```
Or in more hackish multiline way:
```less
paper-drawer-panel {
  -:~";--paper-drawer-panel-left-drawer: {";
    background-color: red;
    color: blue;
  -:~";}";
}
```
---
As for attempts like `%(~"%d", @ruleset);` - No, Less variables are not macros, they are not expanded regardless of context, and specifically for unnamed rulesets the only proper syntax is `@ruleset()` in a ruleset body and *not* in a property value context.

**Comment by Jamtis** [comment]
Well I agree with you that non-standard syntax should generally not be supported.

However this means that LESS in incompatible with Polymer in many cases.

Thanks for the reply anyway. :+1:

**Comment by donny-dont** [comment]
Not sure if this makes a difference in your decision @seven-phases-max but there is an open spec for an `@apply` rule, http://tabatkins.github.io/specs/css-apply-rule/#issue-882293bd.

Also see https://github.com/Polymer/polymer/issues/1373 as it sounds like something Google is going to really push through the process.

**Comment by seven-phases-max** [comment]
Well, Tab Atkins generated *tens* of CSS standard proposals (named "ideas") and while Google guys may push this into their browser even tomorrow, this is still too far from becoming even CSS draft thing (remember CSS `vars` not entering CSS world for about ten years?).

>SASS seems to be addressing this sass/sass#1128

It's not really Sass adressing this, but Tab Atkins pushes Sass ideas (yes, slightly reworked) to W3C.

**Comment by donny-dont** [comment]
@seven-phases-max apologies for the implication. I just wanted to make a note that Polymer is pushing the mixin proposal and Chrome is likely to implement it once its shaped up which was something @Jamtis didn't really communicate.

I'm fairly new to using LESS is there any sort of plugin system for the parser in these sorts of scenarios? Something like an experimental flag you can turn on and off?

**Comment by seven-phases-max** [comment]
I'm reopening this to be able to redirect duplicate requests here.

**Comment by dantman** [comment]
We have a spec for this; a polyfill exists; Chrome has implemented it; Polymer is already using it in a 1.0 release; and websites are already using Polymer.

I think it's a perfectly reasonable expectation for Less to support the syntax at a bare minimum level.

Less already passes `@apply`, `--foo: bar;`, and `var(--foo)` through. The only thing missing is getting Less to pass through the `{ ... }` in `--foo: { property: value; }` with the same basic processing given to other css blocks; instead of coming to a hard stop and throwing a parse error.

**Comment by stramel** [comment]
I think the full spec of CSS Variables should be supported by LESS. Would love for this to happen! Opens

**Comment by matthew-dean** [comment]
@dantman 
> We have a spec for this; a polyfill exists; Chrome has implemented it; Polymer is already using it in a 1.0 release; and websites are already using Polymer.

To me, that's still not a very compelling reason. That translates to: Google, Google, and Google are using this. There's been no interest pop up anywhere else, and there are no indications there ever will be.

@donny-dont 
> I'm fairly new to using LESS is there any sort of plugin system for the parser in these sorts of scenarios? Something like an experimental flag you can turn on and off?

There is a plugin system, but no direct support for changing parsing. So, that would not be a trivial thing to do.

**Comment by dantman** [comment]
>> @dantman 
>> We have a spec for this; a polyfill exists; Chrome has implemented it; Polymer is already using it in a 1.0 release; and websites are already using Polymer.
>
> To me, that's still not a very compelling reason. That translates to: Google, Google, and Google are using this. There's been no interest pop up anywhere else, and there are no indications there ever will be.

That conclusion would make sense if we were talking about a more proprietary and unsupported syntax that deviated significantly from normal css; but we're talking about Less generating parse errors when Less already understands nesting and the syntax in question fits perfectly within CSS' bracket matching based parse rules and doesn't break surrounding css, and the other major CSS preprocessor and postprocessor either are working on supporting it or already work.

**Comment by matthew-dean** [comment]
> That conclusion would make sense if we were talking about a more proprietary and unsupported syntax that deviated significantly from normal css; but we're talking about Less generating parse errors when Less already understands nesting and the syntax in question fits perfectly within CSS' bracket matching based parse rules and doesn't break surrounding css, and the other major CSS preprocessor and postprocessor either are working on supporting it or already work.

Not irrelevant. It's pretty similar to Less's detached rulesets. And there's precedent for other "pass-thru" stuff added to Less. So I'm not necessarily against it. Just skeptical if there is value beyond Polymer at the moment.

**Comment by stramel** [comment]
Official Support for CSS Variables is listed here: http://caniuse.com/#feat=css-variables

Minimum Browser version for default CSS Variable Support
- Chrome 49
- Chrome for Android 51
- Android Browser 51
- Firefox for Android 47
- Firefox 31
- Safari 9.1
- iOS Safari 9.3
- Opera 36
- Opera Mobile 37

With multiple CSS Variable Polyfills out there, browsers implementing it and large browser providers pushing for it. It seems like this would be a valuable feature to have. 

Official Spec: https://drafts.csswg.org/css-variables/

**Comment by matthew-dean** [comment]
@stramel That has nothing to do with this issue. CSS Variables are already supported in Less. The request is about `@apply`, which is not part of that spec.

**Comment by stramel** [comment]
@matthew-dean Sorry, I misunderstood. Regardless, :+1: for CSS Mixins/Custom Property Sets

**Comment by donny-dont** [comment]
@matthew-dean what is the bar for less supporting? If another browser implemented then good to go?

**Comment by matthew-dean** [comment]
@donny-dont Since Less is a community project, there's no specific bar. But yes, something beyond a single browser would make it a more general-purpose feature.

**Comment by henkrijneveld** [comment]
Just if someone is searching this (as I did): you can put all the Polymer non-standard css in a separate file and import this with the inline keyword in your less file. The less file has to interface between less variables and polymer custom properties. But that is the way Polymer wants it, anyway. Agree you cannot implement every deviation from standards.

**Comment by matthew-dean** [comment]
Just to revisit this, I think I unfortunately had a narrow view of the scope of the problem.

That is: yes, `@apply` is just a proposal, BUT REGARDLESS, _this_ is _not_ invalid CSS. Not anymore.
```css
paper-drawer-panel {
  --paper-drawer-panel-left-drawer: {
    background-color: red;
  };
}
```
My mistake was thinking that `@apply` was proposing an extension to the Custom Property syntax, but it isn't. The above _IS_ a valid custom property value, and, as such, should be added to Less support. I wish I had taken the time to dig deeper into the spec. (Here: https://www.w3.org/TR/css-variables/#syntax.)

Less has a core goal to support valid CSS, and Custom Property syntax is implemented fully in every browser as of now. The spec is extremely permissive, and may require some big changes in the way property values are parsed. You can put almost anything into it. Just how crazy you can get with values and whether browsers will allow that, I'm not sure. But, the way I read the spec, mostly anything is valid if it's well formed. Meaning, until you encounter a "top-level" semicolon token, you can go nuts, wrapping all sorts of stuff in curly braces or parentheses.

So 👍 to implementing ASAP.

**Comment by matthew-dean** [comment]
Related: https://github.com/less/less.js/issues/2986

**Comment by matthew-dean** [comment]
I'm marking this both as bug and feature request because a) Less claims to parse custom property values, but fails sometimes (but it's not clear when), but b) the types of values Less needs to handle is beyond the original design of even CSS itself, let alone Less, so it is a new feature.

**Comment by matthew-dean** [comment]
Another test that can be added to Less tests for custom property parsing. This is perfectly valid CSS as of today.
```
.test {
  --custom-function: () => { let x = 1; window['NAMESPACE'].doSomething(x); };
}
```
Essentially, once something is in `()` `[]` or `{}`, you can dump in any characters you want. So, semi-colons would pass in the above example. It requires some recursive parsing until braces get matched and closed (if/when they're opened) before you can test for a closing semi-colon.

I started that work here but not sure yet if this is correct. https://github.com/less/less.js/blob/edge/lib/less/parser/parser.js#L1326

**Comment by seven-phases-max** [comment]
>Essentially, once something is in () [] or {}, you can dump in any characters you want.

mm, is it? For me https://www.w3.org/TR/css-variables/#syntax sounds like any *token* (except those listed there) can be anywhere there (regardless of either kind of parens). Or does this assumed by some other paragraph) Could you, please point me, where it says about *any* character, or anything specific about values *inside* parens? (unless it's `url(....)` of course)?

I.e. are either of `--var: ?;`,  `--var: (?)` or  `--var: [¾]` valid?

**Comment by matthew-dean** [comment]
As to this:

> are either of --var: ?;, --var: (?); or --var: [¾]; valid?

I don't know.

Should we treat them as valid? IMO yes. Because we don't / can't know, but we _can_ successfully parse it without too much trouble.

**Comment by seven-phases-max** [comment]
As for possible impementation... It's not a big problem to allow *anything* there until `;` but then we still should handle at list strings, urls and nested `{}` (that may contain `;` inside). Obviously this would mean *no Less code* within.
A step further would to threat it (at the parsing stage) as a sort of DR with just more tokens allowed (though it looks ike a dead-way since  you can't reuse DR-parser for something like `--var: (1 > 2) whatever {whatever}; )
And finally for something more I can't honestly see anything but writing a ful-featured CSS-tokenizer with some of the less features (both: tokens and then eval) allowed in). A biiiiiiiig problem in other words :(

**Comment by seven-phases-max** [comment]
>Only the top-level semicolon-token matters. I interpret that to mean "not within some other matched pair"? But I could be misinterpreting what "top-level" means.

Yes I'm pretty sure it's about the terminating `;`. I.e. as in: `--var: ";" url(;) {;};` is valid and `--var ; {} foo;` is not.

**Comment by matthew-dean** [comment]
> Obviously this would mean no Less code within.

My only thought is that we could maaaaybe allow interpolation syntax, in case you wanted to generate custom properties. But maybe that's sort of a "round 2 if there's interest" kind of idea. 

> I'm not sure about just (;) though.

I'm not either? I believe specs often have parsing details published somewhere (like token path diagrams), but I'm not sure if it exists in this case. As a stage 1 - only requiring matching `{}`, `()`, and quotes, and dumping anything into an anonymous node until a top-level `;` seems fine. Matching brackets `[]` may not be necessary, but they're also mentioned and it's fairly trivial once all the other pieces are there.

I don't think we need to specially tokenize anything, or get into mixing this with DR stuff. After all, someone could end up with:
```less
.weird {
  --php: ($x = 5 /* + 15 */ + 5; echo $x;);
  --example: [My DR will be --this: { 
    blah: nope;
    --never mind i gave up;
    no wait here it --is: {
      lol: cats;
    }
  }];
}
```
In other words, it would be better for Less to wipe its hands clean of dealing with the contents of custom properties.

**Comment by matthew-dean** [comment]
Note that if we modify the above example, THIS should be considered invalid:
```less
.weird {
  --example: [My DR will be --this: { 
    blah: nope;
    --never mind i gave up;
    no wait here it --is: {
      lol: cats;
   // missing matching }
  }];
}
```

**Comment by seven-phases-max** [comment]
>"no Less code"

Probably, but this means a downgrade - like now one can write `--var: darken(red, 5%)`; and it works, but then (for the sake of `--fortran: read (*, *, iostat=ierr) radius, height;`) it won't :(

**Comment by matthew-dean** [comment]
> Probably, but this means a downgrade - like now one can write --var: darken(red, 5%); and it works, but then (for the sake of --fortran: read (*, *, iostat=ierr) radius, height;) it won't :(

🤔  Well..... yes I see what you mean, which is why maybe interpolation of some kind might be necessary? We could basically at the parser treat things similar to:
```less
@iostat: ierr;
--fortran: read (*, *, iostat=@{iostat}) radius, height;

// treat as:
--fortran: ~"read (*, *, iostat=@{iostat}) radius, height";

```
(With the aforementioned caveats of proper matching tokens?)

I mean.... the alternative is that we essentially recommend escaping syntax. Just would be nice for fewer modifications of in-place CSS code.

**Comment by seven-phases-max** [comment]
Yes, interpoation wil do the trick. Though it still would be nice to a DR-ike parsing as an experimental option (I'm pretty sure many will prefer real `--var: darken(red, 5%);` to imaginary `--javascript: 1 = 2;`at least until such custom property hacking becomes widespread :)

In other words, I'm fine with both ways (separately or simultaneously).

**Comment by matthew-dean** [comment]
I think I have a fairly robust implementation / solution for this. I used a number of code examples from this thread in tests. So, custom property values and unknown at-rules are essentially treated as escaped quoted values to allow interpolation. Check out #3213.

**Comment by matthew-dean** [comment]
Fixed.
